### ISSUE TITLE
Invalid deployment with Hardhat Network fork option

### ISSUE THREAD

**Issue by abcoathup** [OpenZeppelin/openzeppelin-upgrades#233]
Running tests multiple times or deploying multiple times to a hardhat public network fork (such as mainnet) errors with `Invalid deployment`

An `.openzeppelin/mainnet.json` network file is created for the first test or deploy and then trying again results in the issue.

Reported in the forum: https://forum.openzeppelin.com/t/hardhat-openzeppelin-upgrades-error-invalid-deployment-with-address-and-txhash/4672

It can be reproduced as follows:

1. `Box.sol` contract, `Box.proxy.js` and `deploy.js` from: https://forum.openzeppelin.com/t/openzeppelin-upgrades-step-by-step-tutorial-for-hardhat/3580

## hardhat.config.js
Configure Hardhat fork 
```js
// hardhat.config.js
require("@nomiclabs/hardhat-ethers");
require('@openzeppelin/hardhat-upgrades');

const { alchemyApiKey, mnemonic } = require('./secrets.json');

/**
 * @type import('hardhat/config').HardhatUserConfig
 */
module.exports = {
  solidity: "0.7.3",
  networks: {
    hardhat: {
      forking: {
       url: `https://eth-mainnet.alchemyapi.io/v2/${alchemyApiKey}`
      }
    }
  }
};
``` 

2. Run the test multiple times
```
abcoathup@OpenZeppelin:~/projects/forum/drbullock$ npx hardhat test


  Box (proxy)
    ✓ retrieve returns a value previously initialized


  1 passing (8s)

abcoathup@OpenZeppelin:~/projects/forum/drbullock$ npx hardhat test


  Box (proxy)
    1) "before each" hook for "retrieve returns a value previously initialized"


  0 passing (4s)
  1 failing

  1) Box (proxy)
       "before each" hook for "retrieve returns a value previously initialized":
     Error: Invalid deployment with address 0x5FbDB2315678afecb367f032d93F642f64180aa3 and txHash 0x0aa91d2a882ebd60c10819762c83ed4298214460ae2d20d07d7e7bf6186129ef
      at Object.resumeOrDeploy (node_modules/@openzeppelin/upgrades-core/src/deployment.ts:35:13)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
      at /home/abcoathup/projects/forum/drbullock/node_modules/@openzeppelin/upgrades-core/src/impl-store.ts:33:23
      at Manifest.lockedRun (node_modules/@openzeppelin/upgrades-core/src/manifest.ts:90:14)
      at fetchOrDeployGeneric (node_modules/@openzeppelin/upgrades-core/src/impl-store.ts:25:24)
      at Proxy.deployProxy (node_modules/@openzeppelin/hardhat-upgrades/src/deploy-proxy.ts:45:18)
      at Context.<anonymous> (test/Box.proxy.js:12:11)
```

Otherwise run deploy twice:
```
abcoathup@OpenZeppelin:~/projects/forum/drbullock$ npx hardhat run --network hardhat scripts/deploy.js
Deploying Box...
Box deployed to: 0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0
abcoathup@OpenZeppelin:~/projects/forum/drbullock$ npx hardhat run --network hardhat scripts/deploy.js
Deploying Box...
InvalidDeployment [Error]: Invalid deployment with address 0x5FbDB2315678afecb367f032d93F642f64180aa3 and txHash 0x0aa91d2a882ebd60c10819762c83ed4298214460ae2d20d07d7e7bf6186129ef
    at Object.resumeOrDeploy (/home/abcoathup/projects/forum/drbullock/node_modules/@openzeppelin/upgrades-core/src/deployment.ts:35:13)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at /home/abcoathup/projects/forum/drbullock/node_modules/@openzeppelin/upgrades-core/src/impl-store.ts:33:23
    at Manifest.lockedRun (/home/abcoathup/projects/forum/drbullock/node_modules/@openzeppelin/upgrades-core/src/manifest.ts:90:14)
    at fetchOrDeployGeneric (/home/abcoathup/projects/forum/drbullock/node_modules/@openzeppelin/upgrades-core/src/impl-store.ts:25:24)
    at Proxy.deployProxy (/home/abcoathup/projects/forum/drbullock/node_modules/@openzeppelin/hardhat-upgrades/src/deploy-proxy.ts:45:18)
    at main (/home/abcoathup/projects/forum/drbullock/scripts/deploy.js:5:17) {
  deployment: {
    address: '0x5FbDB2315678afecb367f032d93F642f64180aa3',
    txHash: '0x0aa91d2a882ebd60c10819762c83ed4298214460ae2d20d07d7e7bf6186129ef',
    layout: { storage: [Array], types: [Object] }
  }
}
```

**Comment by frangio** [comment]
The issue here is that the first run stores the deployment in the network file that corresponds to the real network even though it only ran on a fork. The subsequent run will verify that the deployment it found can be reused by checking that it is on chain. If it's not found on chain it fails, because we assume something has gone wrong and we want to let the user know about it.

Will think about how to fix this.

**Comment by snowkidind** [comment]
On a deploy, if you set the gas price too low, an internal timeout occurs ( not sure if hardhat or oz ) 

Something catches, ceasing execution, causing the other two contracts in wait to not get deployed. So your mainnet.jason file never gets updated, causing an error like this. 

Not sure if you can, because of admin linking but it would be nice if there was a 'resume process' functionality that catches these timeouts or deploy errors and attempts to repair them.

**Comment by frangio** [comment]
@snowkidind Is this related to the fork option?

The plugins are already "resumable" if a timeout happens: the tx hash is stored in mainnet.json and subsequent runs will pick it up and wait until it's mined. Did you run into a situation when this wasn't happening correctly? Please open a new issue if so.

**Comment by snowkidind** [comment]
This doesnt relate to the fork, but it does relate to the process stopping. I will add an issue if i observe it again.

**Comment by santisiri** [comment]
Hi, I was getting this same error whole trying to deploy on `kovan` 

```
Deploying contracts with the account: 0xEdB47F691a77022f03817f80FEc09aCc363747aE
Account balance: 1916847260000000000
Deploying UBI Coin...

Warning: Potentially unsafe deployment of UBI

    You are using the `unsafeAllowCustomTypes` flag to skip storage checks for structs and enums.
    Make sure you have manually checked the storage layout for incompatibilities.

InvalidDeployment [Error]: Invalid deployment with address 0x2bdAb779cde50998293de98087Efefd440CC6036 and txHash 0xd811c01067524feb68ea452ce8b08dbb96656e5c58d10f134be46c2717e7d835
    at Object.waitAndValidateDeployment (/Users/santisiri/Ethereum/ubi/node_modules/@openzeppelin/upgrades-core/src/deployment.ts:66:15)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at fetchOrDeployGeneric (/Users/santisiri/Ethereum/ubi/node_modules/@openzeppelin/upgrades-core/src/impl-store.ts:42:5)
    at Proxy.deployProxy (/Users/santisiri/Ethereum/ubi/node_modules/@openzeppelin/hardhat-upgrades/src/deploy-proxy.ts:52:26)
    at main (/Users/santisiri/Ethereum/ubi/scripts/deploy.js:17:17) {
  deployment: {
    address: '0x2bdAb779cde50998293de98087Efefd440CC6036',
    txHash: '0xd811c01067524feb68ea452ce8b08dbb96656e5c58d10f134be46c2717e7d835'
  }
}
```

Was able to push the contract through using `gasMultiplier: 3` in the settings.

**Comment by frangio** [comment]
@santisiri Can you share if you were using Infura? Someone reported something similar [in the forum](https://forum.openzeppelin.com/t/invalid-deployment-using-infura/5371).

**Comment by frangio** [comment]
@santisiri I should have fixed the error you saw in https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/282. However, it's unrelated to the fork option which this issue is about.

**Comment by frangio** [comment]
@abcoathup Can this be reproduced in the latest core? #282 may have fixed this for forked networks as well.

**Comment by abcoathup** [comment]
Hi @frangio,

I wasn't able to reproduce with the latest version of core.  Closing as Fixed by #282
